### DISCUSSION TITLE
tutorials/voyage-part1/09-connecting-data-using-entities

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#61] [course feedback]
# tutorials/voyage-part1/09-connecting-data-using-entities

Learn how to use entities to connect data types between subgraphs

https://www.apollographql.com/tutorials/voyage-part1/09-connecting-data-using-entities

**Comment by edwardyun12** [comment]
Woah. That's a lot of entities :0

**Reply by mabuyo** [reply]
Get ready to hear/read that word some more! :)

**Comment by Dark417** [comment]
Hi M, can you explain:
"An entity can have more than one primary key"
"we can use the @key directive to specify which field(s) can uniquely identify an object of that type"
If AEntity has two fields, both of which have @key directives on them, does gql use both keys to identify one instance, or either one is sufficient to identify?

**Reply by lizhennessy** [reply]
Hi there @Dark417 , this is an excellent question. It's definitely possible for an entity to have more than one key. _Which_ key is used to resolve the entity depends on the subgraph that we're querying data from. Here's an example from the Apollo docs:

> For example, a Reviews subgraph might refer to products by their ID, whereas an Inventory subgraph might use SKUs.

```
type Product @key(fields: "id") @key(fields: "sku") {
  id: ID!
  sku: String!
  name: String!
  price: Int
}
```

So in this case, we can imagine a query for products that requests recent reviews, along with the amount of inventory in stock. This query would need to consult both our imagined `reviews` and `inventory` subgraphs. In this case, both subgraphs use different identifiers; so in looking up the correct reviews for a particular product, our `reviews` subgraph might use the `id` field, whereas the `inventory` subgraph relies on `sku` instead.

You asked whether GraphQL uses _both_ keys to identify one instance: for that to occur, we'd be looking at [compound keys](https://www.apollographql.com/docs/federation/entities-advanced/#compound-keys), which use a combination of two or more keys to resolve a particular instance.

If you're interested in reading more about this or other federation topics, please feel free to check out the [Apollo docs on entities](https://www.apollographql.com/docs/federation/entities-advanced/#multiple-keys)! But of course, let us know if you have any other questions we can help with.

**Comment by BiroGates** [comment]
>*Any subgraph that contributes fields to an entity needs to define a reference resolver function for that entity* 
This does not seems to be right to me. isn't only the owner service that needs to define the __resolverReference ?
